### PR TITLE
Work on foreign keys for nested list mapping

### PR DESCRIPTION
**Pull request by jakobbraun**
* Added key support to mapping language schema
* Added validation for key property
* Extracted interface `ColumnMappingDefinition` from  `AbstractColumnMappingDefinition`
* Improved code readability in `TableMappingFactory` 
* Added key extraction to `TableMappingFactory`

### CODE REVIEW

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractColumnMappingDefinition.java [hunk 1]
```diff
@@ -1,31 +1,26 @@
...
 /**
- * Definition of a column mapping from DynamoDB table to Exasol Virtual Schema.
+ * This abstract class is the common basis for the {@link ColumnMappingDefinition}s.
  * <p>
- * Each instance of this class represents one column in the Exasol table. Objects of this class get serialized into the
- * column adapter notes. They are created using a {@link MappingDefinitionFactory}. Storing the Mapping definition is
- * necessary as mapping definition files in BucketFS could be changed, but the mapping must not be changed until a
- * {@code REFRESH} statement is called.
+ * Objects of this class get serialized into the column adapter notes. They are created using a
+ * {@link MappingDefinitionFactory}. Storing the Mapping definition is necessary as mapping definition files in BucketFS
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractColumnMappingDefinition.java:11]:
```suggestion
 * {@link MappingDefinitionFactory}. Storing the mapping definition is necessary as mapping definition files in BucketFS
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractColumnMappingDefinition.java [hunk 2]
```diff
@@ -34,90 +29,39 @@
...
      */
     public static class ConstructorParameters {
         private final String exasolColumnName;
         private final DocumentPathExpression pathToSourceProperty;
         private final LookupFailBehaviour lookupFailBehaviour;
 
         /**
-         * Creates a parameter object for
-         * {@link AbstractColumnMappingDefinition#AbstractColumnMappingDefinition(ConstructorParameters)}
+         * Creates a parameter object for {@link AbstractColumnMappingDefinition(ConstructorParameters )}.
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractColumnMappingDefinition.java:64]:
```suggestion
         * Creates a parameter object for {@link AbstractColumnMappingDefinition(ConstructorParameters)}.
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java [hunk 3]
```diff
@@ -41,73 +25,140 @@ protected AbstractTableMappingFactory() {
...
+        private static final String FIELDS_KEY = "fields";
+        private static final String TO_STRING_MAPPING_KEY = "toStringMapping";
+        private static final String TO_JSON_MAPPING_KEY = "toJsonMapping";
+        private static final String TO_TABLE_MAPPING_KEY = "toTableMapping";
+
+        /**
+         * Building tables for nested lists is delayed using this queue as they need the completely built table that
+         * maps the object they are nested in.
+         */
+        private final List<NestedTableReader> nestedTableReaderQueue;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:62]: The instance name `nestedTableReaderQueue` suggests it behaves as a queue (FIFO), but the data structure is a List. 

Would it make sense to rename the variable name, or change it to Queue interface and LinkedList?

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java [hunk 4]
```diff
@@ -41,73 +25,140 @@ protected AbstractTableMappingFactory() {
...
+            union.addAll(this.keyColumns);
+            union.addAll(this.nonKeyColumns);
+            return union;
+        }
+
+        public List<NestedTableReader> getNestedTableReaderQueue() {
+            return this.nestedTableReaderQueue;
+        }
+
+        public boolean isHasNestedTable() {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:160]:
```suggestion
        public boolean isNestedTable() {
```

Comment (morazow) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:160]: You also have abstract method above with same name, remove it?

File: src/main/java/com/exasol/adapter/dynamodb/mapping/ColumnMappingDefinition.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Get the default value of this column.
+     *
+     * @return {@link ValueExpression} holding default value
+     */
+    public ValueExpression getExasolDefaultValue();
+
+    /**
+     * Get the string representation of the exasol column default value literal.
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/ColumnMappingDefinition.java:37]:
```suggestion
     * Get the string representation of the Exasol column default value literal.
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/ColumnMappingDefinitionKeyTypeReader.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package com.exasol.adapter.dynamodb.mapping;
+
+import javax.json.JsonObject;
+
+/**
+ * This class reads the {@code key} property of a exasol document mapping language column mapping definition.
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/ColumnMappingDefinitionKeyTypeReader.java:6]:
```suggestion
 * This class reads the {@code key} property of an Exasol document mapping language column mapping definition.
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/ColumnMappingDefinitionKeyTypeReader.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ */
+public class ColumnMappingDefinitionKeyTypeReader {
+    private static final String KEY_KEY = "key";
+    private static final String KEY_LOCAL = "local";
+    private static final String KEY_GLOBAL = "global";
+
+    /**
+     * Reads the {@link KeyType} of a column mapping definition.
+     *
+     * @param definition the exasol document mapping language definition of the column
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/ColumnMappingDefinitionKeyTypeReader.java:16]:
```suggestion
     * @param definition the Exasol document mapping language definition of the column
```

File: src/main/java/com/exasol/adapter/dynamodb/mapping/TableMappingDefinition.java [hunk 8]
```diff
@@ -16,20 +16,20 @@
     private static final long serialVersionUID = 3568807256753213582L;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/TableMappingDefinition.java:16]: This also should be updated. Please also update it in all updated classes that has serial uid.

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java [hunk 9]
```diff
@@ -41,73 +25,140 @@ protected AbstractTableMappingFactory() {
...
+                final DocumentPathExpression.Builder sourcePath) {
+            final ColumnMappingDefinitionKeyTypeReader.KeyType columnsKeyType = new ColumnMappingDefinitionKeyTypeReader()
+                    .readKeyType(definition);
+            if (columnsKeyType.equals(ColumnMappingDefinitionKeyTypeReader.KeyType.NO_KEY)) {
+                this.nonKeyColumns.add(column);
+            } else {
+                if (this.keyType != columnsKeyType
+                        && this.keyType != ColumnMappingDefinitionKeyTypeReader.KeyType.NO_KEY) {
+                    throw new ExasolDocumentMappingLanguageException(sourcePath.build().toString()
+                            + ": This table already has a key of different type (global/local).  "
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:113]:
```suggestion
                            + ": This table already has a key of different type (global/local). "
```

File: src/test/java/com/exasol/adapter/dynamodb/mapping/JsonMappingFactoryTest.java [hunk 10]
```diff
@@ -107,4 +105,18 @@ void testException() throws IOException {
...
+        final File invalidFile = this.mappingTestFiles.generateInvalidFile(MappingTestFiles.BASIC_MAPPING_FILE,
+                base -> {
+                    base.getJSONObject("mapping").getJSONObject("fields").getJSONObject("name")
+                            .getJSONObject("toStringMapping").put("key", "local");
+                    return base;
+                });
+        final ExasolDocumentMappingLanguageException exception = assertThrows(
+                ExasolDocumentMappingLanguageException.class, () -> getMappingDefinitionForFile(invalidFile));
+        assertThat(exception.getMessage(), startsWith(
+                "/name: This table already has a key of different type (global/local).  Please either define all keys of the table local or global."));
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/dynamodb/mapping/JsonMappingFactoryTest.java:120]:
```suggestion
                "/name: This table already has a key of different type (global/local). Please either define all keys of the table local or global."));
```

File: src/main/java/com/exasol/adapter/dynamodb/remotetablequery/ColumnLiteralComparisonPredicate.java [hunk 11]
```diff
@@ -10,7 +10,7 @@
 public class ColumnLiteralComparisonPredicate<DocumentVisitorType> extends ComparisonPredicate<DocumentVisitorType> {
     private static final long serialVersionUID = -5861703862819874391L;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/dynamodb/remotetablequery/ColumnLiteralComparisonPredicate.java:11]: Update serial uid.

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java (hunk 4 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:160]: These are two different methods. This one is in the nested class and should be `hasNestedTable()` the other one is in the outer class and is called `isNestedTable()`

File: src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java (hunk 3 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/dynamodb/mapping/AbstractTableMappingFactory.java:62]: yes ;-)

**Pull request merged**